h/services/oauth.py

```diff
@@ -161,10 +144,51 @@ def create_token(self, user, authclient):
 
         return token
 
-    def _decode(self, token, **kwargs):
+
+class GrantToken(object):
+    """
+    Represents a JWT bearer grant token.
+
+    This class is responsible for a couple of things: firstly, verifying that
+    the token is a correctly-formatted JSON Web Token, and that it contains all
+    the required claims in the right formats. Some of this processing is
+    deferred to the `jwt` module, but that doesn't handle all the fields we
+    want to validate.
+
+    """
+
+    def __init__(self, token):
+        if not token or type(token) != text_type:
+            raise OAuthTokenError('required assertion parameter is missing',
+                                  'invalid_request')
+        self._token = token
+
+        try:
+            self._claims = jwt.decode(token, verify=False)
+        except jwt.DecodeError:
+            raise OAuthTokenError('invalid JWT signature', 'invalid_grant')
+
+    @property
+    def issuer(self):
+        iss = self._claims.get('iss', None)
+        if not iss:
+            raise OAuthTokenError('grant token issuer is missing', 'invalid_grant')
+        return iss
+
+    @property
+    def subject(self):
```

Comment:
That’s an excellent point. I had an idea about a potentially neat way to address it (bbf4360); let me know what you think.